server/src/main/java/com/thoughtworks/go/server/service/AgentService.java

```diff
@@ -126,143 +140,409 @@ public AgentsViewModel registeredAgents() {
+                .collect(Collectors.toList());
+    }
+
+    private List<String> getUUIDsToRemoveEnvFrom(List<String> uuidsToAssociateWithEnv, List<String> uuidsAssociatedWithEnv) {
+        return uuidsAssociatedWithEnv.stream()
+                .filter(uuid -> !uuidsToAssociateWithEnv.contains(uuid))
+                .collect(Collectors.toList());
+    }
+
+    private List<Agent> getAgentsToAddEnvToList(List<String> uuidsToAddEnvsTo, String env) {
+        return uuidsToAddEnvsTo.stream()
+                .map(uuid -> {
+                    Agent agent = agentDao.agentByUuid(uuid);
+                    String envsToSet = append(agent.getEnvironments(), singletonList(env));
+                    agent.setEnvironments(envsToSet);
+                    return agent;
+                })
+                .collect(Collectors.toList());
+    }
+
+    private List<Agent> getAgentsToRemoveEnvFromList(List<String> agentsToRemoveEnvFrom, String env) {
+        return agentsToRemoveEnvFrom.stream()
+                .map(uuid -> {
+                    Agent agent = agentDao.agentByUuid(uuid);
+                    String envsToSet = remove(agent.getEnvironments(), singletonList(env));
+                    agent.setEnvironments(envsToSet);
+                    return agent;
+                })
+                .collect(Collectors.toList());
+    }
+
+    private List<String> getAssociatedUUIDs(EnvironmentAgentsConfig asssociatedAgents) {
+        List<String> associatedUUIDs = emptyList();
+
+        if (!isEmpty(asssociatedAgents)) {
+            associatedUUIDs = asssociatedAgents.stream().map(EnvironmentAgentConfig::getUuid).collect(Collectors.toList());
+        }
+
+        return associatedUUIDs;
+    }
+
+    private HashMap<String, AgentConfigStatus> createAgentToStatusMap(List<Agent> agents) {
+        HashMap<String, AgentConfigStatus> agentToStatusMap = new HashMap<>();
+
+        agents.forEach(agent -> {
+            AgentInstance agentInstance = agentInstances.findAgent(agent.getUuid());
+            AgentConfigStatus agentConfigStatus = agentInstance.getStatus().getConfigStatus();
+            agentToStatusMap.put(agent.getUuid(), agentConfigStatus);
+        });
+
+        return agentToStatusMap;
+    }
+
+    private void enableDisableAgent(TriState enable, Agent agent) {
+        if (enable.isTrue()) {
+            agent.setDisabled(false);
+        } else if (enable.isFalse()) {
+            agent.setDisabled(true);
+        }
+    }
+
+    private void addOnlyThoseEnvsThatAreNotAssociatedWithTheAgentFromConfigRepo(EnvironmentsConfig envsToAdd, Agent agent) {
+        if (envsToAdd != null) {
+            String uuid = agent.getUuid();
+            envsToAdd.forEach(env -> {
+                if (env.containsAgentRemotely(uuid)) {
+                    LOGGER.info(format("Not adding Agent [%s] to Environment [%s] as it is already associated from a Config Repo", uuid, env.name().toString()));
```

Comment:
I think it might be useful to have this logged. May be as debug log? Any specific reason to remove it?

Reply:
It feels odd for me to have these kind of information logged and would lead to clutter. 

I would probably add a note in the API documentation to explain the behaviour, I think it's very unlikely for a user to check the logs to see why an agent environment is not set(well its already set to the environment, but from PaC) .

Reply:
I have removed the logging.